src/workerd/api/streams/writable.c++

```diff
@@ -280,6 +282,137 @@ jsg::Ref<WritableStream> WritableStream::constructor(
+  kj::Promise<void> waitForCompletionOrRevoke() {
+    auto paf = kj::newPromiseAndFulfiller<void>();
+    doneFulfiller = kj::mv(paf.fulfiller);
+
+    return paf.promise.attach([weakRef = weakRef->addRef()]() mutable {
+      KJ_IF_SOME(obj, weakRef->tryGet()) {
+        // Stream is still alive, revoke it.
+        if (!obj.canceler.isEmpty()) {
+          obj.canceler.cancel(cancellationException());
+        }
+        obj.inner = kj::none;
+      }
+    });
+  }
+
+  kj::Promise<void> write(const void* buffer, size_t size) override {
+    return canceler.wrap(getInner().write(buffer, size));
+  }
+  kj::Promise<void> write(kj::ArrayPtr<const kj::ArrayPtr<const byte>> pieces) override {
+    return canceler.wrap(getInner().write(pieces));
+  }
+
+  // TODO(perf): We can't properly implement tryPumpFrom(), which means that Cap'n Proto will
+  //   be unable to perform path shortening if the underlying stream turns out to be another capnp
+  //   stream. This isn't a huge deal, but might be nice to enable someday. It may require
+  //   significant refactoring of streams.
+
+  kj::Promise<void> whenWriteDisconnected() override {
+    // TODO(someday): WritableStreamSink doesn't give us a way to implement this.
+    return kj::NEVER_DONE;
+  }
+
+  kj::Promise<void> end() override {
+    return canceler.wrap(getInner().end());
+  }
+
+private:
+  kj::Maybe<kj::Own<WritableStreamSink>> inner;
+  kj::Canceler canceler;
+  kj::Own<kj::PromiseFulfiller<void>> doneFulfiller;
+  kj::Own<WeakRef<WritableStreamRpcAdapter>> weakRef =
+      kj::refcounted<WeakRef<WritableStreamRpcAdapter>>(
+          kj::Badge<WritableStreamRpcAdapter>(), *this);
+
+  WritableStreamSink& getInner() {
+    return *KJ_UNWRAP_OR(inner, {
+      kj::throwFatalException(cancellationException());
+    });
+  }
+
+  static kj::Exception cancellationException() {
+    return JSG_KJ_EXCEPTION(DISCONNECTED, Error,
+        "WritableStream received over RPC was disconnected because the remote execution context "
+        "has endeded.");
+  }
+};
+
+}  // namespace
+
+void WritableStream::serialize(jsg::Lock& js, jsg::Serializer& serializer) {
+  // Serialize by effectively creating a `JsRpcStub` around this object and serializing that.
+  // Except we don't actually want to do _exactly_ that, because we do not want to actually create
+  // a `JsRpcStub` locally. So do the important parts of `JsRpcStub::constructor()` followed by
+  // `JsRpcStub::serialize()`.
+
+  auto& handler = JSG_REQUIRE_NONNULL(serializer.getExternalHandler(), DOMDataCloneError,
+      "WritableStream can only be serialized for RPC.");
+  auto externalHandler = dynamic_cast<RpcSerializerExternalHander*>(&handler);
+  JSG_REQUIRE(externalHandler != nullptr, DOMDataCloneError,
+      "WritableStream can only be serialized for RPC.");
+
+  IoContext& ioctx = IoContext::current();
+
+  // TODO(soon): Support JS-backed WritableStreams. Currently this only supports native streams
```

Comment:
@jasnell Is there an easy way to support this? Basically I need a `WritableStreamSink` representing the `WritableStream` so I can easily adapt it to a KJ stream. I think there are two options, when the `WritableStream` is not backed by a `WritableStreamSink`:

1. Create an `IdentityTransformStreamImpl`, which is always backed by native source/sink. Use the writable side of that, and separately pump the readable side into the JS stream.
2. Create an alternate version of `WritableStreamRpcAdapter`, above, which locks the isolate on every call and then calls into the JS Writer API.

Is there a better option?

Reply:
For short term, I would probably just have a `WritableStreamSink` implementation that wraps the `WritableStream` and writes directly to it's underlying controller, acquiring the isolate lock on every write. Using the `write(...)` that passes `kj::ArrayPtr<kj::Array<...>> pieces` would end up being the most performant way to write if you have multiple chunks. Longer term I'd like to explore more efficient ways of optimizing these kind of KJ-to-JS and JS-to-KJ pipelines for both this and the `PumpToAdapter` in JS-backed ReadableStream